File: Carrier.Android.SDK/app/src/main/java/org/elastos/carrier/CarrierExtension.java
package org.elastos.carrier;

import org.elastos.carrier.exceptions.CarrierException;

public abstract class CarrierExtension {
    private static final String TAG = "CarrierExtension";
    private Carrier carrier;
    private long nativeCookie = 0;

    private native boolean native_init(Carrier carrier);
    private native boolean invite_friend(Carrier carrier, String to, String data,
                                         FriendInviteResponseHandler handler);
    private native boolean reply_friend(Carrier carrier, String to, int status, String reason, String data);
    private native TurnServerInfo get_turn_server(Carrier carrier);
    private native void native_cleanup(Carrier carrier);
    private static native int get_error_code();

    protected static class TurnServerInfo {
        private String server;
        private String username;
        private String password;
        private String realm;
        private int port;

        TurnServerInfo(String server, String username, String password, String realm,
                       int port) {
            this.server = server;
            this.username = username;
            this.password = password;
            this.realm = realm;
            this.port = port;
        }

        public String getServer() {
            return server;
        }

        public int getPort() {
            return port;
        }

        public String getUsername() {
            return username;
        }

        public String getPassword() {
            return password;
        }

        public String getRealm() {
            return realm;
        }
    }

    abstract protected void onFriendInvite(Carrier carrier, String from, String data);

    protected CarrierExtension(Carrier carrier) {
        if (carrier == null)
            throw new IllegalArgumentException();

        this.carrier = carrier;

        Log.i(TAG, "CarrierExtension instance created");
    }

    protected TurnServerInfo getTurnServerInfo() throws CarrierException {
        TurnServerInfo sinfo;

        if (nativeCookie == 0)
            throw new IllegalStateException();

        sinfo = get_turn_server(carrier);
        if (sinfo == null)
            throw CarrierException.fromErrorCode(get_error_code());

        Log.d(TAG, String.format("Get TURN server info {\n  server: %s\n  username: %s\n  password: %s\n" +
                        "  realm: %s\n  port: %d\n}",
              sinfo.getServer(), sinfo.getUsername(), sinfo.getPassword(), sinfo.getRealm(), sinfo.getPort()));

        return sinfo;
    }

    protected void inviteFriend(String to, String data, FriendInviteResponseHandler handler)
            throws CarrierException {
        if (nativeCookie == 0)
            throw new IllegalStateException();

        if (to == null || to.length() == 0 ||
                data == null || data.length() == 0 || handler == null)
            throw new IllegalArgumentException();

        Log.d(TAG, "Inviting friend " + to + "with greet data " + data);

        if (!invite_friend(carrier, to, data, handler))
            throw CarrierException.fromErrorCode(get_error_code());

        Log.d(TAG, "Send friend invite request to " + to);
    }

    protected void replyFriendInvite(String to, int status, String reason, String data)
            throws CarrierException {
        if (nativeCookie == 0)
            throw new IllegalStateException();

        if (to == null || to.length() == 0 || (status != 0 && reason == null))
            throw new IllegalArgumentException();

        if (status == 0)
            Log.d(TAG, String.format("Attempt to confirm friend invite to %s with data [%s]",
                    to, data));
        else
            Log.d(TAG, String.format("Attempt to refuse friend invite to %s with status %d," +
                    "and reason %s", to, status, reason));

        if (!reply_friend(carrier, to, status, reason, data))
            throw CarrierException.fromErrorCode(get_error_code());

        if (status == 0)
            Log.d(TAG, String.format("Confirmed friend invite to %s with data [%s]", to, data));
        else
            Log.d(TAG, String.format("Refused friend invite to %s with status %d and " +
                    "reason %s", to, status, reason));
    }

    protected void registerExtension() throws CarrierException {
        if (nativeCookie != 0)
            return;

        if (!native_init(carrier))
            throw CarrierException.fromErrorCode(get_error_code());
    }

    @Override
    protected void finalize() throws Throwable {
        if (nativeCookie != 0)
            native_cleanup(carrier);
        super.finalize();
    }
}

File: Carrier.Swift.SDK/ElastosCarrier/Carrier/CarrierExtension.swift

import Foundation

@objc(ELACarrierExtension)
open class CarrierExtension: NSObject {

    public typealias CarrierExtensionInviteReplyCallback =
        (_ carrier: Carrier, _ from: String, _ status: Int, _ reason: String?,
        _ data: String?) -> Void

    public typealias CarrierExtensionInviteCallback =
        (_ carrier: Carrier, _ from: String, _ data: String?) -> Void

    private let TAG = "CarrierExtension"
    private var carrier: Carrier?
    private var handler: CarrierExtensionInviteCallback?
    private var didCleanup: Bool

    deinit {
        cleanup()
    }

    public init(_ carrier: Carrier) {
        self.carrier = carrier
        Log.i(TAG, "CarrierExtension instance created")
        self.didCleanup = true
        super.init()
    }

    @objc (turnServerInfo:)
    public func turnServerInfo() throws -> TurnServerInfo {
        var cinfo = CTurnServer()
        let result = ela_get_turn_server(carrier!.ccarrier!, &cinfo)
        guard result >= 0 else {
            let errno: Int = getErrorCode()
            Log.e(TAG, "Get turn server error: 0x%X", errno)
            throw CarrierError.FromErrorCode(errno: errno)
        }

        return TurnServerInfo.convertCTurnServerToTurnServerInfo(cinfo)
    }

    @objc(sendInviteFriendRequest:data:responseHandler:error:)
    public func sendInviteFriendRequest(to target: String, withData data: String, _ responseHandler: @escaping CarrierExtensionInviteReplyCallback) throws {
        guard !target.isEmpty else {
            throw CarrierError.InvalidArgument
        }
        Log.d(TAG, "Inviting friend " + target + "with greet data " + data)
        let cb: CExtensionInviteReplyCallback = {

            (_, cfrom, cstatus, creason, cdata, clen, cctxt) in

            let ectxt = Unmanaged<AnyObject>.fromOpaque(cctxt)
                .takeRetainedValue() as! [AnyObject?]

            let carrier = ectxt[0] as! Carrier
            let handler = ectxt[1] as! CarrierExtensionInviteReplyCallback

            let from = String(cString: cfrom)
            let status = Int(cstatus)
            var reason: String? = nil
            var _data : String? = nil

            if status != 0 {
                reason = String(cString: creason!)
            } else {
                _data = String(cString: cdata!)
            }

            handler(carrier, from, status, reason, _data)
        }

        let econtext: [AnyObject?] = [carrier, responseHandler as AnyObject]
        let unmanaged = Unmanaged.passRetained(econtext as AnyObject)
        let cctxt = unmanaged.toOpaque()
        
        let result = target.withCString { (cto) -> Int32 in
            return data.withCString { (cdata) -> Int32 in
                let len = data.utf8CString.count
                return extension_invite_friend(carrier!.ccarrier!, cto, cdata, len, cb, cctxt)
            }
        }
        guard result >= 0 else {
            unmanaged.release()
            let errno = getErrorCode()
            Log.e(TAG, "Invite friend to \(target) error: 0x%X", errno)
            throw CarrierError.FromErrorCode(errno: errno)
        }

        Log.d(TAG, "Sended friend invite request to \(target).")
    }

    @objc(replyFriendInviteRequest:status:resson:data:error:)
    public func replyFriendInviteRequest(to target: String, withStatus status: Int, _ reason: String?, _ data: String?) throws {
        guard !target.isEmpty else {
            throw CarrierError.InvalidArgument
        }
        if status == 0 {
            Log.d(TAG, "Attempt to confirm friend invite to " + target + "with data \(data)")
        }
        else {
            Log.d(TAG, "Attempt to confirm friend invite to " + target + "with status \(data)" + "and reason \(reason)")
        }

        var creason: UnsafeMutablePointer<Int8>?
        var cdata: UnsafeMutablePointer<Int8>?
        var len: Int = 0

        let result = target.withCString { (cto) -> Int32 in
            if status != 0 {
                creason = reason!.withCString() { (ptr) in
                    return strdup(ptr)
                }
            } else {
                cdata = data!.withCString() { (ptr) in
                    return strdup(ptr)
                }

                len = data!.utf8CString.count
            }

            defer {
                if creason != nil {
                    free(creason)
                }
                if cdata != nil {
                    free(cdata)
                }
            }
            return extension_reply_friend_invite(carrier!.ccarrier!, cto, CInt(status), creason, cdata, len)
        }
        guard result >= 0 else {
            let errno = getErrorCode()
            throw CarrierError.FromErrorCode(errno: errno)
        }

        if status == 0 {
            Log.d(TAG, "Confirmed friend invite to " + target + "with data \(data)")
        }
        else {
            Log.d(TAG, "Refused friend invite to " + target + "with status \(data)" + "and reason \(reason)")
        }
    }

    @objc (registerExtension:error:)
    public func registerExtension(_ handler: @escaping CarrierExtensionInviteCallback) throws {

        let carrierExtension = CarrierExtension(carrier!)
        carrierExtension.handler = handler
        let cb: CExtensionInviteCallback = {

            (_, cfrom, cdata, clen, cctxt) in

            let ccarrierExtension = Unmanaged<CarrierExtension>.fromOpaque(cctxt!)
                .takeUnretainedValue()

            let carrier = ccarrierExtension.carrier
            let handle = ccarrierExtension.handler

            let from = String(cString: cfrom!)
            let _data : String? = String(cString: cdata)

            handle!(carrier!, from, _data)
        }

        let cctxt = Unmanaged.passRetained(carrierExtension).toOpaque()
        let result = extension_init(carrier!.ccarrier!, cb, cctxt)
        guard result >= 0 else {
            let errno = getErrorCode()
            throw CarrierError.FromErrorCode(errno: errno)
        }

        carrierExtension.didCleanup = false
    }

    @objc (cleanup)
    public func cleanup() {

        objc_sync_enter(self)
        if !didCleanup {
            extension_cleanup(carrier!.ccarrier!)
            carrier = nil
            didCleanup = true
        }
        objc_sync_exit(self)
    }
}

